### ISSUE TITLE
Decide on limits for Client API

### ISSUE THREAD

**Issue by Aaron1011** [SpongePowered/McTester#2]
Currently, the Client API deliberately only allows for running one-off commands or actions on the client. There is no `Client#runOnClient(Runnable runnable)` method that would allow for running arbitrary code on the client thread.

Adding such a method would have the immediate benefit of making tests simpler to write. More complex tests could be written without the need to complicate the `Client` interface.

The downside to adding such a method is that it would permanently rule out support for out-of-process clients. as described in https://github.com/Aaron1011/McTester/issues/1. There is no way to run an arbitrary `Runnable` on another process, so any multi-client support would have to be accomplished through the multi-classloader approach.

**Comment by liach** [comment]
It may be a difficulty. Minecraft itself has a lot of public fields that are hard to restrict their usages. Some of the usages of the fields are not bad, such as player, world, and open gui access.

**Comment by Aaron1011** [comment]
Of course, I can't prevent people from writing tests that rely on the client. However, I'm not sure if I want to officially support it, since it will limit how I can expand McTester.

The nice thing about using the method in Client is that it gives you the synchronization and ordering you would expect - once the method returns, the server  has already processed the client's reply.

**Comment by liach** [comment]
If you can have `runOnClient` method, I suggest you changing it to a consumer of `Minecraft` since that class is one of the most frequently used object on the client
